app/src/swap/handle_sign_transaction.c

```diff
@@ -0,0 +1,187 @@
+#include "bignum.h"
+#include "crypto.h"
+#include "lib_standard_app/swap_lib_calls.h"
+#include "parser.h"
+#include "swap.h"
+#include "swap_utils.h"
+#include "zxformat.h"
+
+#define SWAP_EXPECTED_ITEMS 3
+
+swap_globals_t G_swap_state;
+
+// Save the BSS address where we will write the return value when finished
+static uint8_t *G_swap_sign_return_value_address;
+
+////////////////////////////////////////////////////////////////
+
+bool copy_transaction_parameters(create_transaction_parameters_t *sign_transaction_params) {
+    if (sign_transaction_params == NULL) {
+        return false;
+    }
+
+    // First copy parameters to stack, and then to global data.
+    // We need this "trick" as the input data position can overlap with globals
+    char destination_address[65] = {0};
+    uint8_t amount[16] = {0};
+    uint8_t amount_length = {0};
+    uint8_t fees[8] = {0};
+
+    strncpy(destination_address, sign_transaction_params->destination_address, sizeof(destination_address) - 1);
+
+    if ((destination_address[sizeof(destination_address) - 1] != '\0') || (sign_transaction_params->amount_length > 16) ||
+        (sign_transaction_params->fee_amount_length > 8)) {
+        return false;
+    }
+
+    // store amount as big endian in 16 bytes, so the passed data should be alligned to right
+    // input {0xEE, 0x00, 0xFF} should be stored like {0x00, 0x00, 0x00, 0x00, 0x00, 0xEE, 0x00, 0xFF}
+    memcpy(amount + 16 - sign_transaction_params->amount_length, sign_transaction_params->amount,
+           sign_transaction_params->amount_length);
+
+    memcpy(fees + 8 - sign_transaction_params->fee_amount_length, sign_transaction_params->fee_amount,
+           sign_transaction_params->fee_amount_length);
+
+    amount_length = sign_transaction_params->amount_length;
+
+    // Full reset the global variables
+    os_explicit_zero_BSS_segment();
+
+    // Keep the address at which we'll reply the signing status
+    G_swap_sign_return_value_address = &sign_transaction_params->result;
+
+    // Commit the values read from exchange to the clean global space
+    G_swap_state.amount_length = amount_length;
+    memcpy(G_swap_state.amount, amount, sizeof(amount));
+    memcpy(G_swap_state.destination_address, destination_address, sizeof(G_swap_state.destination_address));
+    readU64BE(fees, &G_swap_state.fees);
+
+    return true;
+}
+
+// Ensure the received transaction matches what was validated in the Exchange app UI
+parser_error_t check_swap_conditions(parser_tx_t *txObj) {
+    parser_error_t err = parser_unexpected_error;
+    if (txObj == NULL) {
+        return err;
+    }
+    // Check transaction method arguments number. Balance transfer Should be 3 (for tx v26).
+    // [dest(address type) |  dest(address) | value(amount)]
+    // We will check that index 5 does not have the TIP
+    if (txObj->blob.totalMethodItems != SWAP_EXPECTED_ITEMS && txObj->blob.totalMethodItems != SWAP_EXPECTED_ITEMS + 1) {
```

Comment:
Could you explain why are we allowing `txObj->blob.totalMethodItems == SWAP_EXPECTED_ITEMS + 1`? What's the "TIP"?

Reply:
Improvement made following slack discussion